substrate-node/runtime/src/migrations/tfgrid_v15_smart_contract_v8.rs

```diff
@@ -0,0 +1,231 @@
+        log::info!(
+            "🔎 MigrateTwinsV15 pre migration: Number of existing twins {:?}",
+            twins_count
+        );
+
+        info!("👥  TFGrid pallet to v14 passes PRE migrate checks ✅",);
+        Ok(twins_count.encode())
+    }
+
+    fn on_runtime_upgrade() -> Weight {
+        let mut twins: BTreeMap<u32, pallet_tfgrid::types::Twin<AccountIdOf<T>>> = BTreeMap::new();
+        let mut total_reads = 0;
+        let mut total_writes = 0;
+        if pallet_tfgrid::PalletVersion::<T>::get()
+            == pallet_tfgrid::types::StorageVersion::V14Struct
+        {
+            let (reads, writes) = migrate_tfgrid::<T>(&mut twins);
+            total_reads += reads;
+            total_writes += writes;
+        } else {
+            info!(" >>> Unused TFGrid pallet V15 migration");
+        }
+        if pallet_smart_contract::PalletVersion::<T>::get()
+            >= pallet_smart_contract::types::StorageVersion::V6
+        {
+            let (reads, writes) = migrate_smart_contract::<T>(&twins);
+            total_reads += reads;
+            total_writes += writes;
+        } else {
+            info!(" >>> Unused Smart Contract pallet V8 migration");
+        }
+
+        if twins.len() > 0 {
+            let (reads, writes) = write_twins::<T>(&twins);
+            total_reads += reads;
+            total_writes += writes;
+
+        }
+        T::DbWeight::get().reads_writes(total_reads, total_writes)
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade(pre_twins_count: Vec<u8>) -> Result<(), &'static str> {
+        info!(
+            "current pallet version: {:?}",
+            pallet_tfgrid::PalletVersion::<T>::get()
+        );
+        assert!(
+            pallet_tfgrid::PalletVersion::<T>::get()
+                >= pallet_tfgrid::types::StorageVersion::V15Struct
+        );
+
+        // Check number of twins against pre-check result
+        let pre_twins_count: u64 = Decode::decode(&mut pre_twins_count.as_slice())
+            .expect("the state parameter should be something that was generated by pre_upgrade");
+        assert_eq!(
+            pallet_tfgrid::Twins::<T>::iter().count() as u64,
+            pre_twins_count,
+            "Number of twins migrated does not match"
+        );
+
+        info!(
+            "👥  TFGrid pallet migration to {:?} passes POST migrate checks ✅",
+            pallet_tfgrid::Pallet::<T>::pallet_version()
+        );
+
+        Ok(())
+    }
+}
+
+pub fn write_twins<T: Config>(twins: &BTreeMap<u32, pallet_tfgrid::types::Twin<AccountIdOf<T>>>) -> (u64, u64) {
+    info!(" >>> Writing twins to storage...");
+    for (twin_id, twin) in twins {
+        pallet_tfgrid::Twins::<T>::insert(&twin_id, &twin);
+    }
+    // Update pallet storage version
+    pallet_tfgrid::PalletVersion::<T>::set(pallet_tfgrid::types::StorageVersion::V15Struct);
+    info!(" <<< Twin migration success, storage version upgraded ({:?} twins)", pallet_tfgrid::Twins::<T>::iter().count());
```

Comment:
remove count in the info log, this log will always run and impact the migration execution duration